### PR TITLE
Type constraint index tuning

### PR DESCRIPTION
**Pull request by hedengran**

### CODE REVIEW

**Comment by neo-technology-commit-status-publisher** [comment]
Thanks for the documentation updates.

The preview documentation has now been torn down - reopening this PR will republish it.

**Review by blouerat — APPROVED** [review]

File: modules/ROOT/pages/query-tuning/indexes.adoc [hunk 1]
```diff
@@ -171,6 +171,11 @@ point.distance(n.prop, center) < = distance
 
 |===
 
+[NOTE]
+====
+The above set of predicate can be extended with the use of type constraints, see xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[Extending index compatibility with type constraints].
```
Comment (JPryce-Aklundh) [inline modules/ROOT/pages/query-tuning/indexes.adoc]:
```suggestion
The above set of predicates can be extended with the use of type constraints.
For more information, see xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[Extending index compatibility with type constraints].
```

File: modules/ROOT/pages/query-tuning/indexes.adoc [hunk 2]
```diff
@@ -233,6 +238,11 @@ CONTAINS
...
```
Comment (JPryce-Aklundh) [inline modules/ROOT/pages/query-tuning/indexes.adoc]:
```suggestion
The above set of predicates can be extended with the use of type constraints, see xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[Extending index compatibility with type constraints].
```

File: modules/ROOT/pages/query-tuning/indexes.adoc [hunk 3]
```diff
@@ -248,7 +258,22 @@ Depending on how values that are not of type string should be treated, there are
 `MATCH (n:Label) WHERE $param STARTS WITH '' AND n.prop = $param`
 * If expressions which are not of type string should be converted to string, then wrapping these in `toString(<expression>)` is the right choice:
 `MATCH (n:Label) WHERE n.prop = toString($param)`
+* If it's known that the property is always of type string, then it's also possible to xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[add a type constraint to help the planner].
```
Comment (JPryce-Aklundh) [inline modules/ROOT/pages/query-tuning/indexes.adoc]:
```suggestion
* If it is known that the property is always of type `STRING`, then it's also possible to xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[add a type constraint to help the planner].
```

File: modules/ROOT/pages/query-tuning/indexes.adoc [hunk 4]
```diff
@@ -248,7 +258,22 @@ Depending on how values that are not of type string should be treated, there are
...
+
+[[extending-index-compatibility-with-type-constraints]]
+== Extending index compatibility with type constraints
+
+As explained in above sections, for indexes that are compatible only with specific types (i.e. `TEXT` and `POINT` indexes) the planner needs to be able to deduce that a predicate will evaluate to `null` for non-compatible values.
```
Comment (JPryce-Aklundh) [inline modules/ROOT/pages/query-tuning/indexes.adoc]:
```suggestion
For indexes that are compatible only with specific types (i.e. `TEXT` and `POINT` indexes), the planner needs to be able to deduce that a predicate will evaluate to `null` for non-compatible values.
```

File: modules/ROOT/pages/query-tuning/indexes.adoc [hunk 5]
```diff
@@ -248,7 +258,22 @@ Depending on how values that are not of type string should be treated, there are
...
+Since xref::constraints/index.adoc#types-of-constraint[type constraints] guarantee that a property is always of the same type, they can be used to extend the scenarios for which `TEXT` indexes and `POINT` indexes are compatible with a predicate.
+
+For example if the property `prop` in below clause has been constrained to have type `String`, then a `TEXT` index can also be planned for the `IS NOT NULL` predicate:
```
Comment (JPryce-Aklundh) [inline modules/ROOT/pages/query-tuning/indexes.adoc]:
```suggestion
For example, if the property `prop` in the below query has been constrained to have type `STRING`, then a `TEXT` index can also be planned for the `IS NOT NULL` predicate:
```

File: modules/ROOT/pages/query-tuning/indexes.adoc [hunk 6]
```diff
@@ -248,7 +258,22 @@ Depending on how values that are not of type string should be treated, there are
...
+
+[source]
+----
+MATCH (n: Label) WHERE n.prop IS NOT NULL
+----
 
+Similarly, if the property had been constrained to have type `Point`, then a `POINT` index could have been used.
```
Comment (JPryce-Aklundh) [inline modules/ROOT/pages/query-tuning/indexes.adoc]:
```suggestion
Similarly, if the property had been constrained to have the type `POINT`, then a `POINT` index could have been used.
```

**Review by JPryce-Aklundh — COMMENTED** [review]
Thanks @hedengran - I have made a few editorial suggestions. I have also suggested some changes to the formating of how we write types. This is in accordance with the work the Cypher surface team is currently doing on types (e.g. https://github.com/neo4j/docs-cypher/pull/678)

**Comment by hedengran** [comment]
> Thanks @hedengran - I have made a few editorial suggestions. I have also suggested some changes to the formating of how we write types. This is in accordance with the work the Cypher surface team is currently doing on types (e.g. #678)

Thank you for the excellent suggestions, they have been added in the latest commit.

File: modules/ROOT/pages/query-tuning/indexes.adoc [hunk 7]
```diff
@@ -248,7 +259,22 @@ Depending on how values that are not of type string should be treated, there are
 `MATCH (n:Label) WHERE $param STARTS WITH '' AND n.prop = $param`
 * If expressions which are not of type string should be converted to string, then wrapping these in `toString(<expression>)` is the right choice:
 `MATCH (n:Label) WHERE n.prop = toString($param)`
+* If it is known that the property is always of type `STRING`, then it's also possible to xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[add a type constraint to help the planner].
+
+[[extending-index-compatibility-with-type-constraints]]
+== Extending index compatibility with type constraints
+
```
Comment (JPryce-Aklundh) [inline modules/ROOT/pages/query-tuning/indexes.adoc:266]:
```suggestion
== Extending index compatibility with type constraints

_This feature was introduced in Neo4j 5.12._
```

File: modules/ROOT/pages/query-tuning/indexes.adoc [hunk 8]
```diff
@@ -233,6 +239,11 @@ CONTAINS
 
 |===
 
+[NOTE]
+====
+The above set of predicates can be extended with the use of type constraints, see xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[Extending index compatibility with type constraints].
```
Comment (JPryce-Aklundh) [inline modules/ROOT/pages/query-tuning/indexes.adoc]:
```suggestion
As of Neo4j 5.12, the above set of predicates can be extended with the use of type constraints, see xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[Extending index compatibility with type constraints].
```

File: modules/ROOT/pages/query-tuning/indexes.adoc [hunk 9]
```diff
@@ -171,6 +171,12 @@ point.distance(n.prop, center) < = distance
 
 |===
 
+[NOTE]
+====
+The above set of predicates can be extended with the use of type constraints.
```
Comment (JPryce-Aklundh) [inline modules/ROOT/pages/query-tuning/indexes.adoc]:
```suggestion
As of Neo4j 5.12, The above set of predicates can be extended with the use of type constraints.
```

**Review by JPryce-Aklundh — APPROVED** [review]
Thanks @hedengran - My apologies, but in my earlier review I missed to add information about from what version this is available.  Could you double check my suggestions, and then either apply them, or I can apply them/merge.

File: modules/ROOT/pages/query-tuning/indexes.adoc (hunk 9 above)
Comment (hedengran) [reply modules/ROOT/pages/query-tuning/indexes.adoc]:
```suggestion
As of Neo4j 5.11, the above set of predicates can be extended with the use of type constraints.
```

File: modules/ROOT/pages/query-tuning/indexes.adoc (hunk 8 above)
Comment (hedengran) [inline modules/ROOT/pages/query-tuning/indexes.adoc]:
```suggestion
As of Neo4j 5.11, the above set of predicates can be extended with the use of type constraints, see xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[Extending index compatibility with type constraints].
```

File: modules/ROOT/pages/query-tuning/indexes.adoc [hunk 10]
```diff
@@ -248,7 +259,22 @@ Depending on how values that are not of type string should be treated, there are
 `MATCH (n:Label) WHERE $param STARTS WITH '' AND n.prop = $param`
 * If expressions which are not of type string should be converted to string, then wrapping these in `toString(<expression>)` is the right choice:
 `MATCH (n:Label) WHERE n.prop = toString($param)`
+* If it is known that the property is always of type `STRING`, then it's also possible to xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[add a type constraint to help the planner].
+
+[[extending-index-compatibility-with-type-constraints]]
+== Extending index compatibility with type constraints
```
Comment (hedengran) [inline modules/ROOT/pages/query-tuning/indexes.adoc:265]:
```suggestion
== Extending index compatibility with type constraints

_This feature was introduced in Neo4j 5.11._

```

**Pull request merged**